src/protocols/secure_channel/CASESession.cpp

```diff
@@ -1099,91 +1104,181 @@ CHIP_ERROR CASESession::HandleSigma2(System::PacketBufferHandle && msg)
     return err;
 }
 
-CHIP_ERROR CASESession::SendSigma3()
+struct CASESession::SendSigma3Work
 {
-    MATTER_TRACE_EVENT_SCOPE("SendSigma3", "CASESession");
-    CHIP_ERROR err = CHIP_NO_ERROR;
-
-    MutableByteSpan messageDigestSpan(mMessageDigest);
-    System::PacketBufferHandle msg_R3;
-    size_t data_len;
+    // Status of background processing.
+    CHIP_ERROR status;
 
-    chip::Platform::ScopedMemoryBuffer<uint8_t> msg_R3_Encrypted;
-    size_t msg_r3_encrypted_len;
+    // Session to use after background processing.
+    CASESession * session;
 
-    uint8_t msg_salt[kIPKSize + kSHA256_Hash_Length];
+    // Sequence number used to coordinate foreground/background work for a
+    // particular session establishment.
+    int sequence;
```

Comment:
If I am reading the code correctly, I think this comment is misleading.  Foreground / background coordination seems to be provided with the `CASESession * session` pointer in `SendSigma3Work`.

The sequence number only seems to be used for auditing (asserts).  Are there any expected cases where `CASESession` / `SendSigma3Work` will get out of sequence?

Reply:
Hi Michael, thanks for reviewing! It's more complicated than that. Yes there is a pointer to the appropriate CASESession, and if it were unique it would be sufficient. Alas, the other classes do pairing with a singleton instance of CASESession, so it's not unique. (When we change this, we're going to have a few places in a few classes we need to fix, not just here.) Because scheduled background work may not execute for an arbitrary amount of time, the sequence number ensures that it is occurring during the appropriate pairing session (else it is ignored, as if it did not execute). This is the same mechanism in HandleSigma3c.

Reply:
Ah, I see.  I think that makes sense.

> Because scheduled background work may not execute for an arbitrary amount of time, the sequence number ensures that it is occurring during the appropriate pairing session

Would the reuse of the `CaseSession` instance occur because the `CASESession` instance times  out and gives up on the background work, or because the calling application decides to use the `CASESession` instance for something else before pairing is complete?  If it is the latter, would that have even been safe to do before this change?  I had assumed that the `CASESession` object was off limits until some callback occurs.

Reply:
This then worries me.  What if we aren't using singleton-like `CASESession` objects and `CASESession` times out and is deleted, but then the work object still dereferences its `CASESession` object pointer after that?

Reply:
That is true for the `CASEServer`, but isn't `Sigma3` sent from the initiator?  It is quite feasible with the current SDK to initiate many concurrent instances of CASE with a dynamically allocated `CASESession` objects.

I think my question really comes down to this: as initiator, is it guaranteed to be safe to delete a `CASESession` object after `SessionEstablishmentDelegate::OnSessionEstablished` or `SessionEstablishmentDelegate::OnSessionEstablishmentError` is called, or is there a possibility that the background work could still access the `CASESession` object after this?

Reply:
Ah, good point. Sigh. Is there one SDK place where initiators start CASE sessions that I could check, or could it be multiple places?

Reply:
Here is where they live:

https://github.com/project-chip/connectedhomeip/blob/379cde1b9c1051ab162c77d188c1e4a7cf7943ca/src/app/CASEClientPool.h#L46

And here is where they are allocated:

https://github.com/project-chip/connectedhomeip/blob/379cde1b9c1051ab162c77d188c1e4a7cf7943ca/src/app/CASESessionManager.cpp#L49

Reply:
I finally bit the bullet and added a template helper for managing these tasks. (Been thinking of doing it more like this for a while, but was trying to avoid complicating things if not necessary, but now it's helpful.) So the helper holds on to the work, and ensures it gets cleaned up via reference counting, which the session also participates in. The session also cleans up the pointer in the work back to the session, if it's cleared/aborted/etc. That should take care of it, without having to change the lifetimes of sessions themselves.

Reply:
OK, that addresses my concern then.  Thanks!

Reply:
Yeah thanks for pointing out that while the responder side uses a singleton for pairing, the initiator side doesn't necessarily.